File: pec-observation-app/android-app/app/src/main/kotlin/br/gov/educacao/sp/pecobservation/adapters/ui/viewmodels/SyncStatusViewModel.kt
package br.gov.educacao.sp.pecobservation.adapters.ui.viewmodels

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.gov.educacao.sp.pecobservation.adapters.sync.SyncWorker
import br.gov.educacao.sp.pecobservation.application.usecases.SyncUseCase
import br.gov.educacao.sp.pecobservation.ports.SyncQueuePort
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class SyncStatusViewModel @Inject constructor(
    private val syncQueue: SyncQueuePort,
    private val syncUseCase: SyncUseCase,
    @ApplicationContext private val context: Context,
) : ViewModel() {

    private val _pendingCount = MutableStateFlow(0)
    val pendingCount: StateFlow<Int> = _pendingCount.asStateFlow()

    private val _isSyncing = MutableStateFlow(false)
    val isSyncing: StateFlow<Boolean> = _isSyncing.asStateFlow()

    private val _lastResult = MutableStateFlow<String?>(null)
    val lastResult: StateFlow<String?> = _lastResult.asStateFlow()

    init { refreshCount() }

    fun refreshCount() {
        viewModelScope.launch {
            _pendingCount.value = syncQueue.pendingCount()
        }
    }

    fun syncNow() {
        if (_isSyncing.value) return
        viewModelScope.launch {
            _isSyncing.value = true
            try {
                val result = syncUseCase.execute()
                _lastResult.value = "Sincronizado: ${result.succeeded} ok, ${result.failed} falhas"
            } catch (e: Exception) {
                _lastResult.value = "Erro: ${e.message}"
            } finally {
                _isSyncing.value = false
                refreshCount()
            }
        }
    }

    fun scheduleBackgroundSync() {
        SyncWorker.schedule(context)
    }
}

File: pec-observation-app/ios-app/Sources/PECObservation/Adapters/UI/ViewModels/SyncStatusViewModel.swift
#if os(iOS)
import Foundation
import Observation

@Observable
public final class SyncStatusViewModel {
    private let syncQueue: SyncQueuePort
    private let syncUseCase: SyncUseCase
    public var pendingCount: Int = 0
    public var lastSyncAt: Date?
    public var syncError: String?
    public var isSyncing = false

    public init(syncQueue: SyncQueuePort, syncUseCase: SyncUseCase) {
        self.syncQueue = syncQueue
        self.syncUseCase = syncUseCase
    }

    public func refreshStatus() {
        pendingCount = (try? syncQueue.pendingCount()) ?? 0
    }

    public func syncNow() async {
        guard !isSyncing else { return }
        isSyncing = true
        syncError = nil
        await syncUseCase.processPendingCommands()
        pendingCount = (try? syncQueue.pendingCount()) ?? 0
        lastSyncAt = Date()
        isSyncing = false
    }
}
#endif
